Enhance fakepg: send and recv

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107626
* __->__ #107625

from working on a starter task with @wanchaol (T161350434):
Enhance Fake Process Group by adding missing collectives: send, recv

xmfan committed:
Enhance fakepg: send and recv

[ghstack-poisoned]

fduwjj reviewed (approved):
LGTM

wanchaol reviewed (approved):
lgtm

wanchaol commented on test/distributed/test_fake_pg.py:
```diff
@@ -153,6 +153,21 @@ def test_alltoall_base(self):
+        dist.init_process_group(backend="fake", rank=0, world_size=2, store=store)
+
+        tensor = torch.ones(3, 3)
+        dist.send(tensor, 1)
```
can you add some assertions on the shape of the send call too?

xmfan committed:
Update on "Enhance fakepg: send and recv"


from working on a starter task with wanchaol (T161350434):
Enhance Fake Process Group by adding missing collectives: send, recv


[ghstack-poisoned]

xmfan committed:
Update on "Enhance fakepg: send and recv"


from working on a starter task with wanchaol (T161350434):
Enhance Fake Process Group by adding missing collectives: send, recv

Differential Revision: [D48606653](https://our.internmc.facebook.com/intern/diff/D48606653)

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

xmfan commented:
removing diff reference from PR body

@pytorchbot merge

xmfan commented:
@pytorchbot label "topic: not user facing"

xmfan commented:
@pytorchbot merge

xmfan commented:
unlinked internal diff
@pytorchbot merge
